File: tests/org.eclipse.rap.rwt.test/src/org/eclipse/rwt/widgets/BrowserHistory_Test.java
/*******************************************************************************
 * Copyright (c) 2009, 2012 EclipseSource and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    EclipseSource - initial API and implementation
 ******************************************************************************/
package org.eclipse.rwt.widgets;


import junit.framework.TestCase;

import org.eclipse.rap.rwt.testfixture.Fixture;
import org.eclipse.rap.rwt.testfixture.Message;
import org.eclipse.rap.rwt.testfixture.Message.CallOperation;
import org.eclipse.rwt.IBrowserHistory;
import org.eclipse.rwt.RWT;
import org.eclipse.rwt.events.BrowserHistoryEvent;
import org.eclipse.rwt.events.BrowserHistoryListener;
import org.eclipse.rwt.internal.protocol.ProtocolTestUtil;
import org.eclipse.rwt.lifecycle.PhaseId;
import org.eclipse.rwt.lifecycle.ProcessActionRunner;
import org.eclipse.swt.widgets.Display;
import org.json.JSONArray;
import org.json.JSONException;


public class BrowserHistory_Test extends TestCase {

  private Display display;

  protected void setUp() throws Exception {
    Fixture.setUp();
    display = new Display();
    Fixture.fakeNewRequest( display );
  }

  protected void tearDown() throws Exception {
    Fixture.tearDown();
  }

  public void testCreateEntry() {
    IBrowserHistory history = RWT.getBrowserHistory();
    try {
      history.createEntry( null, "name" );
      fail( "BrowserHistory#mark must not allow id == null" );
    } catch( IllegalArgumentException e ) {
      // expected
    }
    try {
      history.createEntry( "", "name" );
      fail( "BrowserHistory#mark must not id to be an empty string" );
    } catch( IllegalArgumentException e ) {
      // expected
    }
    try {
      history.createEntry( null, null );
      fail( "BrowserHistory#mark must not allow null for name or title" );
    } catch( IllegalArgumentException e ) {
      // expected
    }
  }

  public void testAddBrowserHistoryListener() {
    final IBrowserHistory history = RWT.getBrowserHistory();
    try {
      history.addBrowserHistoryListener( null );
      fail( "BrowserHistory#addBrowserHistoryListener must not allow null" );
    } catch( IllegalArgumentException e ) {
      // expected
    }
  }

  public void testRemoveBrowserHistoryListener() {
    final IBrowserHistory history = RWT.getBrowserHistory();
    try {
      history.removeBrowserHistoryListener( null );
      fail( "BrowserHistory#removeBrowserHistoryListener must not allow null" );
    } catch( IllegalArgumentException e ) {
      // expected
    }
  }

  public void testRenderCreate() {
    RWT.getBrowserHistory();

    Fixture.executeLifeCycleFromServerThread();

    Message message = Fixture.getProtocolMessage();
    assertNotNull( message.findCreateOperation( "bh" ) );
  }

  public void testRenderCreate_OnlyOnce() {
    RWT.getBrowserHistory();

    Fixture.executeLifeCycleFromServerThread();
    Fixture.fakeNewRequest( display );
    Fixture.executeLifeCycleFromServerThread();

    Message message = Fixture.getProtocolMessage();
    assertNull( message.findCreateOperation( "bh" ) );
  }

  public void testRenderAddNavigationListener() {
    Fixture.fakePhase( PhaseId.READ_DATA );
    ProcessActionRunner.add( new Runnable() {
      public void run() {
        RWT.getBrowserHistory().addBrowserHistoryListener( new BrowserHistoryListener() {
          public void navigated( BrowserHistoryEvent event ) {
          }
        } );
      }
    } );

    Fixture.executeLifeCycleFromServerThread();

    Message message = Fixture.getProtocolMessage();
    assertEquals( Boolean.TRUE, message.findListenProperty( "bh", "navigation" ) );
  }

  public void testRenderRemoveNavigationListener() {
    final BrowserHistoryListener listener = new BrowserHistoryListener() {
      public void navigated( BrowserHistoryEvent event ) {
      }
    };
    RWT.getBrowserHistory().addBrowserHistoryListener( listener );
    Fixture.fakePhase( PhaseId.READ_DATA );
    ProcessActionRunner.add( new Runnable() {
      public void run() {
        RWT.getBrowserHistory().removeBrowserHistoryListener( listener );
      }
    } );

    Fixture.executeLifeCycleFromServerThread();

    Message message = Fixture.getProtocolMessage();
    assertEquals( Boolean.FALSE, message.findListenProperty( "bh", "navigation" ) );
  }

  public void testRenderNavigationListenerUnchanged() {
    RWT.getBrowserHistory().addBrowserHistoryListener( new BrowserHistoryListener() {
      public void navigated( BrowserHistoryEvent event ) {
      }
    } );

    Fixture.executeLifeCycleFromServerThread();

    Message message = Fixture.getProtocolMessage();
    assertNull( message.findListenOperation( "bh", "navigation" ) );
  }

  public void testRenderAdd() throws JSONException {
    RWT.getBrowserHistory().createEntry( "testId", "testText" );

    Fixture.executeLifeCycleFromServerThread();

    Message message = Fixture.getProtocolMessage();
    CallOperation operation = message.findCallOperation( "bh", "add" );
    JSONArray entries = ( JSONArray )operation.getProperty( "entries" );
    JSONArray actual1 = entries.getJSONArray( 0 );
    assertTrue( ProtocolTestUtil.jsonEquals( "[\"testId\",\"testText\"]", actual1 ) );
  }

  public void testRenderAdd_NoEntries() {
    RWT.getBrowserHistory().createEntry( "testId", "testText" );

    Fixture.executeLifeCycleFromServerThread();
    Fixture.fakeNewRequest( display );
    Fixture.executeLifeCycleFromServerThread();

    Message message = Fixture.getProtocolMessage();
    assertNull( message.findCallOperation( "bh", "add" ) );
  }

  public void testRenderAddOrder() throws JSONException {
    RWT.getBrowserHistory().createEntry( "testId1", "testText1" );
    RWT.getBrowserHistory().createEntry( "testId2", "testText2" );

    Fixture.executeLifeCycleFromServerThread();

    Message message = Fixture.getProtocolMessage();
    CallOperation operation = message.findCallOperation( "bh", "add" );
    JSONArray entries = ( JSONArray )operation.getProperty( "entries" );
    JSONArray actual1 = entries.getJSONArray( 0 );
    assertTrue( ProtocolTestUtil.jsonEquals( "[\"testId1\",\"testText1\"]", actual1 ) );
    JSONArray actual2 = entries.getJSONArray( 1 );
    assertTrue( ProtocolTestUtil.jsonEquals( "[\"testId2\",\"testText2\"]", actual2 ) );
  }
}

File: tests/org.eclipse.rap.rwt.jstest/js/org/eclipse/rwt/test/tests/BrowserHistoryTest.js
/*******************************************************************************
 * Copyright (c) 2012 EclipseSource and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    EclipseSource - initial API and implementation
 ******************************************************************************/

qx.Class.define( "org.eclipse.rwt.test.tests.BrowserHistoryTest", {

  extend : qx.core.Object,

  members : {

    testCreateBrowserHistoryByProtocol : function() {
      var browserHistory = this._createBrowserHistoryByProtocol();
      assertTrue( browserHistory instanceof rwt.client.History );
      assertFalse( browserHistory._timer.getEnabled() );
    },

    testSetHasNavigationListenerByProtocol : function() {
      var browserHistory = this._createBrowserHistoryByProtocol();
      rwt.protocol.MessageProcessor.processOperation( {
        "target" : "rwt.client.BrowserHistory",
        "action" : "listen",
        "properties" : {
          "Navigation" : true
        }
      } );
      assertTrue( browserHistory._hasNavigationListener );
      assertTrue( browserHistory.hasEventListeners( "request" ) );
      assertTrue( browserHistory._timer.getEnabled() );
      browserHistory.setHasNavigationListener( false );
    },

    testAddByProtocol : function() {
      var browserHistory = this._createBrowserHistoryByProtocol();
      rwt.protocol.MessageProcessor.processOperation( {
        "target" : "rwt.client.BrowserHistory",
        "action" : "call",
        "method" : "add",
        "properties" : {
          "entries" : [ [ "id1", "text1" ] ]
        }
      } );
      assertEquals( "text1", browserHistory._titles[ "id1" ] );
    },

    testSendNavigated : [
      function() {
        var browserHistory = this._createBrowserHistoryByProtocol();
        rwt.protocol.MessageProcessor.processOperation( {
          "target" : "rwt.client.BrowserHistory",
          "action" : "call",
          "method" : "add",
          "properties" : {
            "entries" : [ [ "id1", "text1" ], [ "id2", "text2" ] ]
          }
        } );
        rwt.protocol.MessageProcessor.processOperation( {
          "target" : "rwt.client.BrowserHistory",
          "action" : "listen",
          "properties" : {
            "Navigation" : true
          }
        } );
        org.eclipse.rwt.test.fixture.TestUtil.store( browserHistory );
        org.eclipse.rwt.test.fixture.TestUtil.delayTest( 200 );
      },
      function( browserHistory) {
        browserHistory.__getState = function() { return "id1"; };
        org.eclipse.rwt.test.fixture.TestUtil.forceInterval( browserHistory._timer );

        var message = org.eclipse.rwt.test.fixture.TestUtil.getMessageObject();
        var actual = message.findNotifyProperty( "rwt.client.BrowserHistory",
                                                 "Navigation",
                                                 "entryId" );
        assertEquals( "id1", actual );
      }
    ],

    testSendNavigatedWithUnknownEntry : [
      function() {
        var browserHistory = this._createBrowserHistoryByProtocol();
        rwt.protocol.MessageProcessor.processOperation( {
          "target" : "rwt.client.BrowserHistory",
          "action" : "call",
          "method" : "add",
          "properties" : {
            "entries" : [ [ "id1", "text1" ], [ "id2", "text2" ] ]
          }
        } );
        rwt.protocol.MessageProcessor.processOperation( {
          "target" : "rwt.client.BrowserHistory",
          "action" : "listen",
          "properties" : {
            "Navigation" : true
          }
        } );
        org.eclipse.rwt.test.fixture.TestUtil.store( browserHistory );
        org.eclipse.rwt.test.fixture.TestUtil.delayTest( 200 );
      },
      function( browserHistory) {
        browserHistory.__getState = function() { return "id3"; };
        org.eclipse.rwt.test.fixture.TestUtil.forceInterval( browserHistory._timer );

        var message = org.eclipse.rwt.test.fixture.TestUtil.getMessageObject();
        var actual = message.findNotifyProperty( "rwt.client.BrowserHistory",
                                                 "Navigation",
                                                 "entryId" );
        assertEquals( "id3", actual );
      }
    ],

    /////////
    // Helper

    _createBrowserHistoryByProtocol : function() {
      return rwt.protocol.ObjectRegistry.getObject( "rwt.client.BrowserHistory" );
    }

  }

} );
